### PR TITLE
feat(single build system): Add CMake to allow switching to cmake + vcpkg instead of scons and other build systems

### PR DESCRIPTION
**Pull request by quyykk**
**This PR switches over our build system to CMake + vcpkg!** No more scons, Makefiles, XCode project and CodeBlocks project! Additionally, it uses vcpkg for dependencies. This means that not only do we not have to maintain the dev64/32 folder anymore, but we can use the same libraries on all three platforms and update them all at once! Adding new libraries also becomes trivial!

The only small asterisk is Steam: We can't use library versions that aren't available inside the Steam runtime. That isn't a problem in practice because with the new sniper runtime libraries should get updated very quickly for Steam. 

Modern IDEs like Visual Studio (and Code) and CLion have integrated CMake support, which means that building the game on Windows now simply involves downloading MinGW and CMake and clicking build in one of those IDEs! But fear not, if you still want to use CodeBlocks you can. Or any other build system for that matter, since CMake supports almost all of them. The needed dependencies on Windows (and on the other two platforms) get automatically downloaded and built when needed as well, no matter what IDE you use. 

Command line building is also very easy: 

```bash
$ cmake --preset linux                # configure
$ cmake --build --preset linux-debug  # build
$ ctest --preset linux-test           # unit test
$ ctest --preset linux-benchmark      # benchmarks
$ ctest --preset linux-integration    # integration tests
```
```powershell
PS> cmake --preset mingw
PS> cmake --build --preset mingw-debug
PS> ctest --preset mingw-test
PS> ctest --preset mingw-benchmark
(no integration tests for Windows yet)
```

On Linux by default the system libraries are used. This means that if your distro doesn't have reasonably up-to-date libraries, compiling ES will fail. However, it is still possible to build ES using a CMake option so that the system libraries from vcpkg are used (just like on MacOS and Windows). Distributing that binary isn't possible unless it is part of AppImage/Flatpak since it depends on newer libraries. This does mean that only distros with newer libraries will be able to update their endless-sky package (like Fedora, Arch, latest non-LTS Ubuntu).

Once you have your project generated, if at any time we say add a new file, then as soon as you pull the new changes and hit compile CMake will realize the project has changed and will add the file for you. Just as good as what we have now basically.

Currently we have (on MacOS) custom scripts that download SDL2 and libmad to avoid crashes on older MacOS versions. This is no longer needed since the dependencies are always built from source and we can build them for older MacOS versions directly. This also guards us if some other library on MacOS where to update and introduce such a crash without us knowing about it immediately (like libmad).

Unfortunately, this PR does break every existing workflow. It requires new applications (cmake, ninja) to be installed as well. I think it's definitely worth it especially since the new process is way easier to setup. And you can start using modern IDEs too! 

Thoughts/Feedback greatly appreciated, as well as testers! Especially from you @tehhowch :wink:. Hopefully you like this PR (if not RIP me).

## Github Actions

I had to rewrite large parts of our current Actions and in the process I made several improvements:

- Caching! All three platforms are now 100% cached for CI builds
- Dependencies are also cached
- I combined the test CI actions into the build ones, so that a build now always invokes the unit tests as well (on all 3 platforms).
- I added a reusable workflow to install any needed dependencies
- I've never seen the memory leaks test be useful, so I removed it. We can add more CI that uses proper sanitizers at a later time.
- There's only one parse test Action instead of 3. I don't really see the need to parse data files on 3 platforms.
- Windows CD uses a recent version of GCC to build, while the Windows CI uses gcc 8 to make sure that we are not breaking anything on that compiler.
- MacOS CI is built on MacOS 12
- I've added a Win32 and Win64 CD Release Action.
- Made the CD release workflow upload the Steam depots that can be directly uploaded to Steam
- Updated every `uses` step to use the latest version
- Added a Windows CI using clang-cl, so that people can use Visual Studio to build the game
- maybe some other stuff that I forgot

Since I'm adding a couple of new jobs, I choose to remove some (mentioned in the list above) so that we don't have a million jobs. The CD release workflow artifact were automatically published here: https://github.com/quyykk/endless-sky/releases/tag/v101. 

## Steam changes

I changed the Steam CD to use cmake (of course), but I also made it so that it is possible to easily build the Steam build locally. For that I used Docker Compose. If you have Docker installed, it's as easy as:

```bash
$ docker-compose run steam-x64 # steam-x86 for the x86 build
```

Build files will get written under `build/steam-x64`. You can also choose the x86 build if you want to. I have also changed the Steam runtime to the newest one, sniper, instead of the ancient scout one!

There isn't any caching yet, ~~I still have to figure out how it should work best~~. From what I can tell, it's not worth it to cache the image, because the Dockerfile simply installs a single package, which is pretty fast anyways.

## Release update changes

Instead of relying on scripts to change the version, I changed it to use CMake instead: CMake generates the correct files with the correct version. This means that to update the version only one line has to be changed and everything else will work automagically: CD builds will include the commit hash and `-alpha`, local builds will include `-alpha` and when we create a release the CD release workflows will simply use the version number without anything else.

## Unit tests & co

I integrated all our tests into CMake's testing framework, CTest:

```bash
$ ctest --preset linux-test        # unit test
$ ctest --preset linux-benchmark   # benchmarks
$ ctest --preset linux-integration # integration tests
```

The integration tests are only available on Linux, but the rest can be used on any platform!

## Testing Done

The CD artifacts have been tested on the following OS. Testers appreciated! Even if the OS has already been tested, or if it is not listed below.

- [ ] Ubuntu 18.04 - **dropped** in favor of native Wayland support
- [x] Ubuntu 20.04
- [x] Ubuntu 22.04
- [x] Fedora 36
- [x] Debian 11
- [ ] Debian 9 - **doesn't work**, glibc is even older than Ubuntu 18 (2.27 vs 2.24). LTS support for Deb9 ended so who cares
- [x] Windows 11
- [x] Windows 10
- [ ] Windows 8.1
- [ ] MacOS 10.9
- [x] MacOS 10.15
- [ ] MacOS 11
- [ ] MacOS 12

If you have Windows 64-bit testing the 32-bit binary is also helpful! I will also spin up VMs for the various Linux distros.

Testers welcome!

## TODO

- [ ] Update the AppVeyor config.

### CODE REVIEW

File: readme-developer.md
Comment (petervdmeer) [reply readme-developer.md:28]: Packages `curl` `zip` `unzip` and `tar` also seem to be required according to vcpkg.

File: readme-developer.md
Comment (petervdmeer) [inline readme-developer.md:31]: typo: Ubuntu

Comment (petervdmeer) [inline readme-developer.md:31]: kitware does look quite official, but isn't cmake.org/download a more appropriate download link?

File: readme-developer.md
Comment (petervdmeer) [inline readme-developer.md:48]: Brecht Sanders looks like a nice person, but I prefer to have download links from an organization that is larger than 1 single person.

**Comment by petervdmeer** [comment]
> * Caching! All three platforms are now 100% cached for CI builds

Does this means that there is a lot less rebuilding also on my local computer? Caching for the Bazel build-system is really nice in that it detects what has really changed and only rebuilds the really changed files. Does CMake also do that?

File: readme-developer.md
Comment (MCOfficer) [reply readme-developer.md:31]: It's official, it's even linked on the download page you mentioned (towards the end, under "Alternative Binary Releases").

I'm still in favour of using the cmake download page, because it also links to other methods of distribution (snap, pip).

File: tests/unit/CMakeLists.txt
Comment (petervdmeer) [inline tests/unit/CMakeLists.txt:2]: Can't we automate this (by just specifying complete folders)?

File: source/CMakeLists.txt
Comment (petervdmeer) [inline source/CMakeLists.txt:1]: Can we automate this (by just providing the folder instead)?

File: readme-developer.md
Comment (MCOfficer) [reply readme-developer.md:48]: For reference, this is the project homepage (which we might want to link to, btw): https://winlibs.com/

Unfortunately the sourceforge project we used thus far doesn't go to gcc 12, so there are little alternatives. the [Mingw-W64 website](https://www.mingw-w64.org/downloads/#mingw-builds) links to [this repo](https://github.com/niXman/mingw-builds-binaries) which is... arguably worse?

File: readme-developer.md
Comment (quyykk) [reply readme-developer.md:31]: To be clear, this specific distribution of MinGW is used by chocolatey, which in turn is used by GitHub and used in their official Windows images. So I'd say it's trustworthy?

File: readme-developer.md
Comment (quyykk) [reply readme-developer.md:48]: To be clear, this distribution is used by chocolately and in turn used by GitHub inside their official Windows images. So pretty trustworthy I'd say

File: tests/unit/CMakeLists.txt
Comment (quyykk) [reply tests/unit/CMakeLists.txt:2]: We can't if we don't want to have to recreate the entire project every time a file is added.

**Comment by quyykk** [comment]
> Does this means that there is a lot less rebuilding also on my local computer?

Well yes and no. Only files that have changed will get recompiled, but that is already the case. Scons only builds files that have changed. CMake does as well.

For CI we need to use ccache which wasn't possible on MacOS due to the use of XCode for building. You can use ccache locally, but unless you're frequently rebuilding the entire game from scratch it doesn't have much effect.

The dependencies that get built are also cached locally, but this is already the case as well since we're not even building them.

**Comment by petervdmeer** [comment]
> For CI we need to use ccache which wasn't possible on MacOS due to the use of XCode for building. You can use ccache locally, but unless you're frequently rebuilding the entire game from scratch it doesn't have much effect.

Right, this is the feature I was looking for (and which I know Bazel also has). I often checkout some branch for testing, and then later checkout another branch that I built earlier. The long build-times to rebuild this another branch should be avoidable by using ccache or Bazel. (Scons seems to decide to rebuild based on timestamps, which results in a lot more rebuilding.)

**Comment by quyykk** [comment]
> (Scons seems to decide to rebuild based on timestamps, which results in a lot more rebuilding.)

[It can use timestamps, but that's not the default behavior (MD5 of the file is)](https://scons.org/doc/production/HTML/scons-user/ch06.html). I also build a lot of branches and I have found that scons only rebuilds the files that have truly changed, so anecdotally I find this to be the case too.

> The long build-times to rebuild this another branch should be avoidable

I'm glad I don't have this problem. A full rebuild takes ~30s on my machine, more if I'm building in release mode because of LTO.

File: steam/Dockerfile
Comment (petervdmeer) [inline steam/Dockerfile:1]: Scout -> Sniper?

File: source/Version.h.in
Comment (tibetiroka) [inline source/Version.h.in:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

**Comment by quyykk** [comment]
I'm really happy with this now. I would appreciate it if people could give their thoughts/feedback/etc.

**Comment by gunqqer** [comment]
A problem I ran into while using this was the lack of documentation that I need to init the vcpkg submodule via git. I'm not sure if I just missed a reference to needing to do this, but if it doesn't exist this informatin should be added to the building from source documentation.

**Comment by quyykk** [comment]
Thanks for trying out! Did you check the readme-developer.txt file? A while back I added detailed instructions that mention first thing to remember to clone the submodule.

**Comment by gunqqer** [comment]
Ah it is somehwere. I didn't read it because of all the files changed, I was going off only what you had in your first comment on this page. Well, as long as we have it documented, that should be good. Will the build instructions on the github wiki be updated with this PR or does it need to be done manually?

**Comment by quyykk** [comment]
It will need to be done manually. But tbh, I think we might as well delete that page, no need to duplicate information.

File: .github/workflows/cd.yaml
Comment (petervdmeer) [inline .github/workflows/cd.yaml:80]: A good practice in software engineering is not to run production code form your own personal home-directory (or someone elses home-directory). This is technically a little bit different, since it is a published repository, but I still prefer to host this action under endless-sky, not under your own personal github site, if we need it in the build-process.

File: source/Version.h.in
Comment (petervdmeer) [reply source/Version.h.in:10]: This review-comment still seems open.

File: .github/workflows/cd.yaml
Comment (quyykk) [reply .github/workflows/cd.yaml:80]: > I think I already asked about the lukka actions below, but also there; could we clone/fork them into endless-sky and run them from our own clone/fork?

I don't really see why this is necessary tbh. We wouldn't get any advantage from doing this, because we would just keeping our fork in sync with upstream anyways. Might as well just use the upstream directly.

> A good practice in software engineering is not to run production code form your own personal home-directory (or someone elses home-directory)

I don't think this analogy holds. Github Actions were made to be reusable, so that people can share their actions with others. I think a better analogy is it's like using an external library (SDL, libmad, libpng).

I wouldn't mind moving my action over to this org.

**Comment by petervdmeer** [comment]
Some changes in this PR are good improvements/bugfixes on their own (regardless of the build-framework chosen):
- Moving readme-developer.md towards docs.
- Moving icons and other files to better locations (like /resources).
- Defining the constexpr in EsUuid in a more specific location.
- Fixing the capitilization in README.md
- Fixes in test_conditionsStore.cpp, test_esuuid.cpp, test_weightedList.cpp.

I wouldn't mind if those improvements are done ahead of this PR in a separate PR (to make this PR a bit more focussed and easier to review). This is not a hard requirement from me, so I suggest to only split off the fixes that are easy to split off.

File: .github/workflows/cd.yaml
Comment (petervdmeer) [reply .github/workflows/cd.yaml]: Please do move over the action; it just looks more professional. (And it is a shame that Lukka is creating much better vcpkg installers than that Microsoft is doing on their official repos.)

**Review by petervdmeer — CHANGES_REQUESTED** [review]
Can we get some focus for this PR? I feel that this PR is just trying to do too many things at once and changing to a single buildsystem is very very valuable, but it also conflicts with any other change that adds or moves files.
I also notice that review-comments just remain open and seem to be lost in the long lists of 236 changes.
Added 0.9.17 milestone, we should get this finished soon to unblock other stuff.

For focus:
- If we want to rename .yaml files to .yml, then just do that in a separate PR. (And rename them all in 1 go.)
- If we want to move icon files around, then just do that in a separate PR as well.
- If we want to remove tests that are not valuable, then just do that in a separate PR. (And if that PR gets rejected, then just keep those tests here.)

We might even want to go for a separate PR that just adds CMake as an _additonal_ build-system first so that people can more easily test it before we switch. We could also add some CI-tests on top of the additional CMake build to ensure that CMake remains working (at the locations where it already works.)

**Review by petervdmeer — APPROVED** [review]
This is now actually reviewable. :) And my concerns seem to be handled. :)

**Comment by quyykk** [comment]
Things left to do:

- [ ] verify that the M1 configuration works
- [ ] dependencies are rebuilt a bit too often, so ideally wait until upstream fixes it
- [ ] the mingw build produces some weird af errors
- [ ] ideally fix missing window decorations on wayland

**Review by RisingLeaf — APPROVED** [review]
Works on my M1 mac

**Comment by petervdmeer** [comment]
> Things left to do:

Do those things all need to be done in this PR? I would prefer to merge this soon, and then switch the various platforms over to CMake one-by-one (so in separate PRs per platform).

> * figure out why M1 macs can't use system libraries

Doesn't need to block this PR, but needs to be fixed before the PR that switches the Apple/Mac build over to CMake.

> * dependencies are rebuilt a bit too often in CI, so ideally wait until upstream fixes it

Depends on the impact; a temporary situation where we move to a single build-system that works well, although with a sub-optimal performance, still sounds better than all contributors always editing files for 5 different build-systems.
If we need to spend a little bit more compute-time, but save this in manual developer work, then the balance still seems reasonable.

> * the mingw build produces some weird af errors

Doesn't need to block this PR, but needs to be fixed before the next PR that switches the Windows build over to CMake.

> * ideally fix missing window decorations on gnome wayland

Does this currently work? If not, then it is not blocking the introduction of cmake.
If it does currently work, then it doesn't need to block this PR, but it would block the PR that switches the Linux build over to CMake.

**Comment by quyykk** [comment]
> Doesn't need to block this PR, but needs to be fixed before the PR that switches the Apple/Mac build over to CMake.

That's true

> If we need to spend a little bit more compute-time, but save this in manual developer work, then the balance still seems reasonable.

inb4 people complain about the long CI times (~~like the one for the integration test~~).

> Doesn't need to block this PR, but needs to be fixed before the next PR that switches the Windows build over to CMake.

I did some investigation, and while I have no idea what causes those errors (and nobody on the internet has the same either), they seem to be harmless. 

> Does this currently work? If not, then it is not blocking the introduction of cmake.
If it does currently work, then it doesn't need to block this PR, but it would block the PR that switches the Linux build over to CMake.

It doesn't that's true. So we keep do this later yeah

**Comment by quyykk** [comment]
This is basically done now. Testing always appreciated :smile: @tehhowch do you have any thoughts/feedback?

**Comment by quyykk** [comment]
One change that I have made in the build instructions that might be slightly controversial is that MinGW is now a second-class citizen for Windows development. To be clear, MinGW works just fine, but developing with MinGW on Windows is really a pain in so many ways that I thought it would be better to recommend using Clang instead.

**Pull request merged**
